### PR TITLE
gh-93274: Expose receiving vectorcall in the Limited API

### LINKED ISSUE #93274: Receiving vector calls in the Py_LIMITED_API
**Feature or enhancement**

The `PyType_FromSpec` family of functions provides a convenient and forward-compatible way mechanism for creating new types. It could in principle also be used to create _callables_ providing a PEP-590 vector call interface, which has significant performance benefits for binding libraries (see the [related discussion](https://discuss.python.org/t/ideas-for-forward-compatible-and-fast-extension-libraries-in-python-3-12/15993/12) here).

One can already specify a member named `__vectorcalloffset__` in `PyType_FromSpec`. This isn't fully working in the limited API, however. I encountered the following problems.

1. The `Py_TPFLAGS_HAVE_VECTORCALL` flag is not part of the limited API.
2. The `PyVectorcall_NARGS()` helper function is not part of the limited API.
3. One would normally set `tp_call` to the compatibility dispatch routine `PyVectorcall_Call`. It is, however, also not part of the public ABI. 
4. Leaving `tp_call` unspecified is not an option. `PyType_Ready()` even throws an exception in `type_ready_pre_checks()` when `tp_call` is unspecified.

**Pitch**

I propose the following changes:

1. Adding ``Py_TPFLAGS_HAVE_VECTORCALL``,  `PyVectorcall_NARGS()`, and `PyVectorcall_Call()` to the limited API.
2. Redundant, but nice: Setting `tp_call` to `PyVectorcall_Call` when `PyType_Ready` encounters a type that doesn't have this field set.

Note that vector calls can be _received_ and _performed_. This issue is just about the receiving end.

**Previous discussion**

See the discord thread https://discuss.python.org/t/ideas-for-forward-compatible-and-fast-extension-libraries-in-python-3-12/15993/12.

### PR DESCRIPTION
**Pull request by encukou**
* Issue: gh-93274

### CODE REVIEW

**Comment by wjakob** [comment]
This looks great. Leaving `tp_call` unchanged sounds fine as long as `PyVectorcall_Call` is available as a stable API-exported symbol.

File: Objects/call.c [hunk 1]
```diff
@@ -1047,3 +1047,11 @@ _PyStack_UnpackDict_Free(PyObject *const *stack, Py_ssize_t nargs,
     PyMem_Free((PyObject **)stack - 1);
     Py_DECREF(kwnames);
 }
+
+// Export for the stable ABI
+#undef PyVectorcall_NARGS(n)
```
Comment (erlend-aasland) [inline Objects/call.c]:
```suggestion
#undef PyVectorcall_NARGS
```

File: Modules/_testcapi/vectorcall_limited.c [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+static PyMethodDef TestMethods[] = {
+    {NULL},
+};
+
+int
+_PyTestCapi_Init_VectorcallLimited(PyObject *m) {
+    if (PyModule_AddFunctions(m, TestMethods) < 0) {
+        return -1;
+    }
+
```
Comment (erlend-aasland) [inline Modules/_testcapi/vectorcall_limited.c]: `TestMethods` appears to be unused; left-overs from early revision?

```suggestion
int
 _PyTestCapi_Init_VectorcallLimited(PyObject *m) {
```

**Review by erlend-aasland — APPROVED** [review]
Looks good, with minor nits!

BTW, in `Modules/_testcapi/vectorcall_limited.c`, everything but `_PyTestCapi_Init_VectorcallLimited` method can be made `static`. Not that it matters much for a test module, though.

File: Modules/_testcapi/vectorcall_limited.c (hunk 2 above)
Comment (encukou) [reply Modules/_testcapi/vectorcall_limited.c]: Left-overs from a “template“, or itself a template/space for future additions. The `PyMethodDef` is usually part of a `PyModuleDef` and using `PyModule_AddFunctions` is unusual, so I prefer to give a hint as to how it's supposed to work.
All that should be in a comment, of course. I'll add one.

**Comment by encukou** [comment]
If anyone needs the TraceRefs buildbot green, please merge the revert. I'm working on a fix, so maybe a revert won't be necessary.

**Comment by vstinner** [comment]
https://buildbot.python.org/all/#builders/484/builds/1920 failed to build:

```
In file included from ./Include/Python.h:44,
                 from ./Modules/_testcapi/parts.h:1,
                 from ./Modules/_testcapi/vectorcall_limited.c:2:
./Include/object.h:62:4: error: #error Py_LIMITED_API is incompatible with Py_TRACE_REFS
   62 | #  error Py_LIMITED_API is incompatible with Py_TRACE_REFS
      |    ^~~~~
make: *** [Makefile:2857: Modules/_testcapi/vectorcall_limited.o] Error 1
```

> If anyone needs the TraceRefs buildbot green, please merge the revert. I'm working on a fix, so maybe a revert won't be necessary.

IMO it's ok if it's broken for a few days, but it would be nice to fix it before next Python (3.12) release.

**Comment by encukou** [comment]
This'll hopefully fix it: https://github.com/python/cpython/pull/95796

**Pull request merged**